test/e2e/network/endpointslice.go

```diff
@@ -299,6 +303,237 @@ var _ = SIGDescribe("EndpointSlice", func() {
 		deleteEndpointSlices(cs, f.Namespace.Name, svc2)
 		expectEndpointsAndSlices(cs, f.Namespace.Name, svc2, []*v1.Pod{pod1, pod2}, 2, 2, true)
 	})
+
+	// Ensure that garbage collection works well when a Service is deleted and
+	// its EndpointSlices are being rapidly updated.
+	// For more context: https://github.com/kubernetes/kubernetes/issues/91287
+	ginkgo.It("should not get in the way of Service owner garbage collection", func() {
+		sharedLabels := map[string]string{"test-garbage-collection": "for-service"}
+
+		deployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(context.TODO(), &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "test-garbage-collection",
+				Labels: sharedLabels,
+			},
+			Spec: appsv1.DeploymentSpec{
+				Selector: &metav1.LabelSelector{MatchLabels: sharedLabels},
+				Replicas: utilpointer.Int32Ptr(1),
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: sharedLabels,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "container1",
+								Image: imageutils.GetE2EImage(imageutils.Nginx),
+								Ports: []v1.ContainerPort{{
+									Name:          "example-name",
+									ContainerPort: int32(3000),
+								}},
+							},
+						},
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+
+		ownerRef := metav1.NewControllerRef(deployment, schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "Deployment"})
+		svc := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "example-int-port",
+				Labels:          sharedLabels,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			},
+			Spec: v1.ServiceSpec{
+				// No need to wait for pods to be ready for this test
+				PublishNotReadyAddresses: true,
+				Selector:                 sharedLabels,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		esWatch, err := cs.DiscoveryV1beta1().EndpointSlices(f.Namespace.Name).Watch(context.TODO(), metav1.ListOptions{LabelSelector: lSelector})
```

Comment:
this is starting a watch with no resourceVersion, which means it gets a synthetic add event for the existing items... is that what you wanted?

Reply:
Good catch, thanks! Should be fixed now.